paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/PartitionMarkDoneProcedure.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.paimon.table.Table;
+
+import org.apache.flink.table.procedure.ProcedureContext;
+import org.apache.flink.table.utils.PartitionPathUtils;
+
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.CoreOptions.METASTORE_PARTITIONED_TABLE;
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_MARK_DONE_ACTION;
+import static org.apache.paimon.utils.ParameterUtils.getPartitionsWithLinkedHashMap;
+import static org.apache.paimon.utils.Preconditions.checkArgument;
+import static org.apache.paimon.utils.Preconditions.checkNotNull;
+
+/**
+ * Partition mark done procedure. Usage:
+ *
+ * <pre><code>
+ *  CALL sys.partition_mark_done('tableId', 'partition1', 'partition2', ...)
+ * </code></pre>
+ */
+public class PartitionMarkDoneProcedure extends ProcedureBase {
+
+    public static final String IDENTIFIER = "partition_mark_done";
+
+    public String[] call(
+            ProcedureContext procedureContext, String tableId, String... partitionStrings)
+            throws Catalog.TableNotExistException {
+        checkArgument(
+                partitionStrings.length > 0,
+                "partition_mark_done procedure must specify partitions.");
+
+        Identifier identifier = Identifier.fromString(tableId);
+        Table table = catalog.getTable(identifier);
+        checkArgument(
+                table instanceof FileStoreTable,
+                "Only FileStoreTable supports partition_mark_done procedure. The table type is '%s'.",
+                table.getClass().getName());
+
+        FileStoreTable fileStoreTable = (FileStoreTable) table;
+        CoreOptions coreOptions = fileStoreTable.coreOptions();
+        Options options = coreOptions.toConfiguration();
+        List<org.apache.paimon.flink.sink.partition.PartitionMarkDoneAction> actions =
+                getPartitionMarkDoneActions(fileStoreTable, coreOptions, options);
+
+        for (LinkedHashMap<String, String> partition :
+                getPartitionsWithLinkedHashMap(partitionStrings)) {
+            try {
+                for (org.apache.paimon.flink.sink.partition.PartitionMarkDoneAction action :
+                        actions) {
+                    action.markDone(PartitionPathUtils.generatePartitionPath(partition));
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return new String[] {"Success"};
+    }
+
+    private List<PartitionMarkDoneAction> getPartitionMarkDoneActions(
+            FileStoreTable fileStoreTable, CoreOptions coreOptions, Options options) {
+        return Arrays.asList(options.get(PARTITION_MARK_DONE_ACTION).split(",")).stream()
+                .map(
+                        action -> {
+                            switch (action) {
+                                case "success-file":
```

Comment:
Extract a static method in `PartitionMarkDoneAction`.

Reply:
done